File: labels/labels.py
"""
Determine the regions of the map and their names.

Input: [{id,x,y,r}], DB.mapskw (id->keywords)
Output: [{x,y,name}]

"""

import math
import argparse
import json
from xiwi.common import mysql

class Paper(object):
    @classmethod
    def make_paper(cls, db_cursor, id, x, y):
        hits = db_cursor.execute('SELECT keywords FROM mapskw WHERE id=%s', (id,))
        if hits != 1:
            keywords = []
        else:
            keywords = db_cursor.fetchone()[0].split(',')
            for i in range(len(keywords)):
                if keywords[i].startswith('Higgs'):
                    keywords[i] = 'Higgs'
        return Paper(id, x, y, keywords)

    def __init__(self, id, x, y, keywords):
        self.id = id
        self.x = x
        self.y = y
        self.keywords = keywords

class Graph(object):
    def __init__(self, papers):
        self.papers = papers
        self.min_x = 0
        self.min_y = 0
        self.max_x = 0
        self.max_y = 0
        for paper in papers:
            self.min_x = min(self.min_x, paper.x)
            self.min_y = min(self.min_y, paper.y)
            self.max_x = max(self.max_x, paper.x)
            self.max_y = max(self.max_y, paper.y)

def keywords_for_area(graph, x, y, r, max_kw):
    """Given a position and radius (in world coordinates), returns the keywords for that area in the map."""
    r = r * r
    hist = {}
    for paper in graph.papers:
        if (paper.x - x)**2 + (paper.y - y)**2 < r:
            for kw in paper.keywords:
                try:
                    hist[kw] += 1
                except KeyError:
                    hist[kw] = 1
    keys = []
    for key in hist:
        if hist[key] > 2:
            keys.append((key, hist[key]))
    keys.sort(lambda a, b: cmp(b[1], a[1]))
    area = []
    for i, k in enumerate(keys):
        if i < max_kw:
            area.append(k[0])

    return area

class GridPoint(object):
    def __init__(self, x, y, valid, kws):
        self.x = x
        self.y = y
        self.valid = valid
        self.kws = kws
        self.region = None

class Region(object):
    def __init__(self, id, x, y, kws):
        self.id = id
        self.x = x
        self.y = y
        self.kws = kws

def determine_regions(graph, hex_rad):
    print('creating hex grid')
    hex_w = hex_rad
    hex_h = int(math.sqrt(3) / 2 * hex_rad)
    grid_w = int((graph.max_x - graph.min_x) / hex_w)
    grid_h = int((graph.max_y - graph.min_y) / hex_h)
    grid = []
    for j in range(grid_h):
        if j % 2 == 0:
            odd = 0
        else:
            odd = 1
        for i in range(odd, grid_w):
            y = graph.min_y + j * hex_h
            x = graph.min_x + (i - 0.5 * odd) * hex_w
            kws = keywords_for_area(graph, x, y, hex_rad, 10)

            valid = 0x00
            # points above
            if j >= 1:
                if odd:
                    valid |= 0x03
                else:
                    if i >= 1:
                        valid |= 0x01
                    if i < grid_w - 1:
                        valid |= 0x02
            # points left/right
            if odd:
                if i >= 2:
                    valid |= 0x04
                if i < grid_w - 1:
                    valid |= 0x08
            else:
                if i >= 1:
                    valid |= 0x04
                if i < grid_w - 1:
                    valid |= 0x08
            # points below
            if j < grid_h - 1:
                if odd:
                    valid |= 0x30
                else:
                    if i >= 1:
                        valid |= 0x10
                    if i < grid_w - 1:
                        valid |= 0x20

            # add grid point
            grid.append(GridPoint(x, y, valid, kws))

    # find seed areas:
    #   a hexagon of 6 areas must have at least 2 kw common in their top 10; that defines the name of the area
    print('finding seed areas')
    regions = []
    for offset, g in enumerate(grid):
        if g.valid == 0x3f:
            kws = get_common_keywords7(grid, grid_w, offset)
            if len(kws) != 0:
                region = Region(len(regions), g.x, g.y, kws)
                regions.append(region)
                grid[offset - grid_w].region = region
                grid[offset - grid_w + 1].region = region
                grid[offset - 1].region = region
                grid[offset].region = region
                grid[offset + 1].region = region
                grid[offset + grid_w - 1].region = region
                grid[offset + grid_w].region = region

    # grow areas
    print('growing areas')
    keep_growing = True
    while keep_growing:
        keep_growing = False
        for offset, g in enumerate(grid):
            if g.region is not None:
                if (g.valid & 0x01):
                    if grow_to(g, grid[offset - grid_w]):
                        keep_growing = True
                if (g.valid & 0x02):
                    if grow_to(g, grid[offset - grid_w + 1]):
                        keep_growing = True
                if (g.valid & 0x04):
                    if grow_to(g, grid[offset - 1]):
                        keep_growing = True
                if (g.valid & 0x08):
                    if grow_to(g, grid[offset + 1]):
                        keep_growing = True
                if (g.valid & 0x10):
                    if grow_to(g, grid[offset + grid_w - 1]):
                        keep_growing = True
                if (g.valid & 0x20):
                    if grow_to(g, grid[offset + grid_w]):
                        keep_growing = True

    return regions

def grow_to(g1, g2):
    if g2.region is not None:
        return False
    for kw in g1.region.kws:
        if kw not in g1.kws:
            return False
    g2.region = g1.region
    return True

def get_common_keywords7(grid, grid_w, offset):
    # get the 7 grid points that make a hexagon
    gs = [grid[offset - grid_w], grid[offset - grid_w + 1], grid[offset - 1], grid[offset], grid[offset + 1], grid[offset + grid_w - 1], grid[offset + grid_w]]
    # check all 7 points are unassigned, and build list of keyword sets
    list_of_kws = []
    for g in gs:
        if g.region is not None:
            return []
        list_of_kws.append(g.kws)
    return get_common_keywords(list_of_kws)

def get_common_keywords(list_of_kws):
    # work out common keywords
    kws = [[kw, True] for kw in list_of_kws[0]]
    for i, g in enumerate(list_of_kws):
        if i > 0:
            for kw in kws:
                if kw[0] not in g:
                    kw[1] = False
    kw2 = []
    for kw in kws:
        if kw[1]:
            kw2.append(kw[0])
    return kw2

def do_work(db_cursor, graph_filename):
    with open(graph_filename) as f:
        papers = json.load(f)
    print('read JSON graph data')
    for i, paper in enumerate(papers):
        if i % 10000 == 0:
            print('made {} papers'.format(i))
        papers[i] = Paper.make_paper(db_cursor, paper[0], paper[1], paper[2])
    graph = Graph(papers)
    print('have graph with {} papers'.format(len(graph.papers)))

    regions = determine_regions(graph, 700)

    json_out = []
    for region in regions:
        json_out.append({'x':region.x, 'y':region.y, 'kws':region.kws})
    print json.dumps(json_out, separators=(',', ':'))

def main():
    # command line arguments
    cmd_parser = argparse.ArgumentParser(description='Convert userdata.')
    cmd_parser.add_argument('--db', metavar='<MySQL database>', help='server name (or localhost) of MySQL database to connect to')
    cmd_parser.add_argument('input', nargs=1, help='input JSON file')
    args = cmd_parser.parse_args()

    # connect to the database
    db_connection = mysql.dbconnect(args.db, 0)
    db_cursor = db_connection.cursor()

    # do the work
    do_work(db_cursor, args.input[0])

    # close database connection
    db_connection.close()

if __name__ == '__main__':
    main()

File: labels/labels.go
package main

import (
    //"io"
    //"io/ioutil"
    "flag"
    "os"
    "bufio"
    "fmt"
    "path/filepath"
    //"net"
    //"net/http"
    //"net/http/fcgi"
    //"strconv"
    //"unicode"
    //"encoding/json"
    //"text/scanner"
    "GoMySQL"
    "runtime"
    "strings"
    //"bytes"
    //"time"
    "math"
    //"math/rand"
    //"crypto/sha1"
    //"crypto/sha256"
    //"compress/gzip"
    //"crypto/aes"
    "sort"
    //"net/smtp"
    "log"
    "xiwi"
)

var flagDB = flag.String("db", "", "MySQL database to connect to")
var flagSkipZones = flag.Bool("skip-tiles", false, "Only generate index file not tiles")
var flagMaxCores = flag.Int("cores",-1,"Max number of system cores to use, default is all of them")

func main() {
    // parse command line options
    flag.Parse()

    if flag.NArg() != 1 {
        log.Fatal("need to specify output directory prefix")
    }

    // connect to the db
    db := xiwi.ConnectToDB(*flagDB)
    if db == nil {
        return
    }
    defer db.Close()

    // read in the graph
    graph := ReadGraph(db)

    // build the quad tree
    graph.BuildQuadTree()

    GenerateAllLabelZones(graph, flag.Arg(0))
}

type Paper struct {
    id          uint
    authors     string
    keywords    string
    x           int
    y           int
    radius      int
    age         float32
    label       string
}

type PaperSortId []*Paper
func (p PaperSortId) Len() int           { return len(p) }
func (p PaperSortId) Less(i, j int) bool { return p[i].id > p[j].id }
func (p PaperSortId) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }


type Graph struct {
    papers  []*Paper
    qt      *QuadTree
    MinX, MinY, MaxX, MaxY int
    BoundsX, BoundsY int
}

func cleanJsonString(input string) string {
    // TODO work out exactly which chars are causing
    // parsing error and blacklist or escape them
    // inplace of this whitelist
    validChars := "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789 -/.,<>"

    output := make([]rune, 0, len(input))

    for _, r := range input {
        if strings.ContainsRune(validChars, r) {
            output = append(output, r)
        }
    }

    return string(output)
}

func getPaperById(papers []*Paper, id uint) *Paper {
    lo := 0
    hi := len(papers) - 1
    for lo <= hi {
        mid := (lo + hi) / 2
        if id == papers[mid].id {
            return papers[mid]
        } else if id < papers[mid].id {
            hi = mid - 1
        } else {
            lo = mid + 1
        }
    }
    return nil
}

func QueryMetaData(db *mysql.Client, papers []*Paper) {
    // execute the query
    err := db.Query("SELECT id,authors FROM meta_data")
    if err != nil {
        fmt.Println("MySQL query error;", err)
        return
    }

    // get result set
    result, err := db.UseResult()
    if err != nil {
        fmt.Println("MySQL use result error;", err)
        return
    }

    // get each row from the result
    for {
        row := result.FetchRow()
        if row == nil {
            break
        }

        var ok bool
        var id uint64
        //var maincat string
        //var allcats string
        var authors string
        if id, ok = row[0].(uint64); !ok { continue }
        //if maincat, ok = row[1].(string); !ok { continue }
        //if allcats, ok = row[2].(string); !ok { continue }
        if row[1] == nil {
            continue
        } else if au, ok := row[1].([]byte); !ok {
            continue
        } else {
            authors = string(au)
        }

        paper := getPaperById(papers, uint(id))
        if paper != nil {
            paper.authors = authors
        }
    }

    db.FreeResult()
}

func QueryPapers(db *mysql.Client) []*Paper {
    // count number of papers
    err := db.Query("SELECT count(id) FROM map_data")
    if err != nil {
        fmt.Println("MySQL query error;", err)
        return nil
    }

    // get result set
    result, err := db.UseResult()
    if err != nil {
        fmt.Println("MySQL use result error;", err)
        return nil
    }
    row := result.FetchRow()
    if row == nil {
        fmt.Println("MySQL didn't return a row")
        return nil
    }

    // get number of papers
    var numPapers int64
    var ok bool
    if numPapers, ok = row[0].(int64); !ok {
        fmt.Println("MySQL didn't return a number")
        return nil
    }
    db.FreeResult()

    // allocate paper array
    papers := make([]*Paper, numPapers)

    // execute the query
    err = db.Query("SELECT map_data.id,map_data.x,map_data.y,map_data.r,keywords.keywords FROM map_data,keywords WHERE map_data.id = keywords.id")
    if err != nil {
        fmt.Println("MySQL query error;", err)
        return nil
    }

    // get result set
    result, err = db.UseResult()
    if err != nil {
        fmt.Println("MySQL use result error;", err)
        return nil
    }

    // get each row from the result
    index := 0
    for {
        row := result.FetchRow()
        if row == nil {
            break
        }

        var ok bool
        var id uint64
        var x, y, r int64
        var keywords []byte
        if id, ok = row[0].(uint64); !ok { continue }
        if x, ok = row[1].(int64); !ok { continue }
        if y, ok = row[2].(int64); !ok { continue }
        if r, ok = row[3].(int64); !ok { continue }
        if keywords, ok = row[4].([]byte); !ok { continue }

        var age float64 = float64(index) / float64(numPapers)
        papers[index] = MakePaper(uint(id), int(x), int(y), int(r), age, string(keywords))
        index += 1
    }

    db.FreeResult()

    if int64(index) != numPapers {
        fmt.Println("could not read all papers from map_data/keywords; wanted", numPapers, "got", index)
        return nil
    }

    // Get keywords

    return papers
}

func MakePaper(id uint, x int, y int, radius int, age float64, keywords string) *Paper {
    paper := new(Paper)
    paper.id = id
    paper.x = x
    paper.y = y
    paper.radius = radius
    paper.age = float32(age)
    paper.keywords = keywords;
    return paper
}

func (paper *Paper) DetermineLabel() {
    // work out author string; maximum 2 authors
    aus := strings.SplitN(paper.authors, ",", 3)
    for i, au := range aus {
        // get the last name
        parts := strings.Split(au, ".")
        aus[i] = parts[len(parts) - 1]
    }
    var auStr string
    if len(aus) <= 1 {
        // 0 or 1 author
        auStr = aus[0] + ","
    } else if len(aus) == 2 {
        // 2 authors
        auStr = aus[0] + "," + aus[1]
    } else {
        // 3 or more authors
        auStr = aus[0] + ",et al."
    }

    // work out keyword string; maximum 2 keywords
    kws := strings.SplitN(paper.keywords, ",", 3)
    var kwStr string
    if len(kws) <= 1 {
        // 0 or 1 keywords
        kwStr = paper.keywords + ","
    } else if len(kws) == 2 {
        // 2 keywords
        kwStr = paper.keywords
    } else {
        // 3 or more keywords
        kwStr = kws[0] + "," + kws[1]
    }

    paper.label = kwStr + "," + auStr
}

func ReadGraph(db *mysql.Client) *Graph {
    graph := new(Graph)

    // load positions from the data base
    graph.papers = QueryPapers(db)
    if graph.papers == nil {
        log.Fatal("could not read papers from db")
    }

    // load extra meta data from the data base
    QueryMetaData(db, graph.papers)

    fmt.Printf("read %v papers from db\n", len(graph.papers))

    // determine labels to use for each paper
    for _, paper := range graph.papers {
        paper.DetermineLabel()
    }

    // work out graph bounding box
    for _, paper := range graph.papers {
        if paper.x - paper.radius < graph.MinX { graph.MinX = paper.x - paper.radius }
        if paper.y - paper.radius < graph.MinY { graph.MinY = paper.y - paper.radius }
        if paper.x + paper.radius > graph.MaxX { graph.MaxX = paper.x + paper.radius }
        if paper.y + paper.radius > graph.MaxY { graph.MaxY = paper.y + paper.radius }
    }

    graph.BoundsX = graph.MaxX - graph.MinX
    graph.BoundsY = graph.MaxY - graph.MinY

    fmt.Printf("graph has %v papers; min=(%v,%v), max=(%v,%v)\n", len(graph.papers), graph.MinX, graph.MinY, graph.MaxX, graph.MaxY)

    return graph
}

type QuadTreeNode struct {
    //Parent          *QuadTreeNode
    //SideLength      int
    Leaf            *Paper
    Q0, Q1, Q2, Q3  *QuadTreeNode
}

type QuadTree struct {
    MinX, MinY, MaxX, MaxY, MaxR  int
    Root                    *QuadTreeNode
}

func QuadTreeInsertPaper(parent *QuadTreeNode, q **QuadTreeNode, paper *Paper, MinX, MinY, MaxX, MaxY int) {
    if *q == nil {
        // hit an empty node; create a new leaf cell and put this paper in it
        *q = new(QuadTreeNode)
        //(*q).Parent = parent
        //(*q).SideLength = MaxX - MinX
        (*q).Leaf = paper

    } else if (*q).Leaf != nil {
        // hit a leaf; turn it into an internal node and re-insert the papers
        oldPaper := (*q).Leaf
        (*q).Leaf = nil
        (*q).Q0 = nil
        (*q).Q1 = nil
        (*q).Q2 = nil
        (*q).Q3 = nil
        QuadTreeInsertPaper(parent, q, oldPaper, MinX, MinY, MaxX, MaxY)
        QuadTreeInsertPaper(parent, q, paper, MinX, MinY, MaxX, MaxY)

    } else {
        // hit an internal node

        // check cell size didn't get too small
        if (MaxX <= MinX + 1 || MaxY <= MinY + 1) {
            log.Println("ERROR: QuadTreeInsertPaper hit minimum cell size")
            return
        }

        // compute the dividing x and y positions
        MidX := (MinX + MaxX) / 2
        MidY := (MinY + MaxY) / 2

        // insert the new paper in the correct cell
        if ((paper.y) < MidY) {
            if ((paper.x) < MidX) {
                QuadTreeInsertPaper(*q, &(*q).Q0, paper, MinX, MinY, MidX, MidY)
            } else {
                QuadTreeInsertPaper(*q, &(*q).Q1, paper, MidX, MinY, MaxX, MidY)
            }
        } else {
            if ((paper.x) < MidX) {
                QuadTreeInsertPaper(*q, &(*q).Q2, paper, MinX, MidY, MidX, MaxY)
            } else {
                QuadTreeInsertPaper(*q, &(*q).Q3, paper, MidX, MidY, MaxX, MaxY)
            }
        }
    }
}

func (graph *Graph) BuildQuadTree() {
    qt := new(QuadTree)

    // if no papers, return
    if len(graph.papers) == 0 {
        return
    }

    // first work out the bounding box of all papers
    qt.MinX = graph.papers[0].x
    qt.MinY = graph.papers[0].y
    qt.MaxX = graph.papers[0].x
    qt.MaxY = graph.papers[0].y
    qt.MaxR = graph.papers[0].radius
    for _, paper := range graph.papers {
        if (paper.x < qt.MinX) { qt.MinX = paper.x; }
        if (paper.y < qt.MinY) { qt.MinY = paper.y; }
        if (paper.x > qt.MaxX) { qt.MaxX = paper.x; }
        if (paper.y > qt.MaxY) { qt.MaxY = paper.y; }
        if (paper.radius > qt.MaxR) { qt.MaxR = paper.radius; }
    }

    // increase the bounding box so it's square
    {
        dx := qt.MaxX - qt.MinX
        dy := qt.MaxY - qt.MinY
        if dx > dy {
            cen_y := (qt.MinY + qt.MaxY) / 2
            qt.MinY = cen_y - dx / 2
            qt.MaxY = cen_y + dx / 2
        } else {
            cen_x := (qt.MinX + qt.MaxX) / 2
            qt.MinX = cen_x - dy / 2
            qt.MaxX = cen_x + dy / 2
        }
    }

    // build the quad tree
    for _, paper := range graph.papers {
        QuadTreeInsertPaper(nil, &qt.Root, paper, qt.MinX, qt.MinY, qt.MaxX, qt.MaxY)
    }

    fmt.Printf("quad tree bounding box: (%v,%v) -- (%v,%v)\n", qt.MinX, qt.MinY, qt.MaxX, qt.MaxY)

    // store the quad tree in the graph object
    graph.qt = qt
}

func (q *QuadTreeNode) ApplyIfWithin(MinX, MinY, MaxX, MaxY int, x, y, rx, ry int, f func(paper *Paper)) {
    if q == nil {
    } else if q.Leaf != nil {
        rx += q.Leaf.radius
        ry += q.Leaf.radius
        if x - rx <= q.Leaf.x && q.Leaf.x <= x + rx && y - ry <= q.Leaf.y && q.Leaf.y <= y + ry {
            f(q.Leaf)
        }
    } else if ((MinX <= x - rx && x - rx <= MaxX) || (MinX <= x + rx && x + rx <= MaxX) || (x - rx <= MinX && x + rx >= MaxX)) &&
              ((MinY <= y - ry && y - ry <= MaxY) || (MinY <= y + ry && y + ry <= MaxY) || (y - ry <= MinY && y + ry >= MaxY)) {
        MidX := (MinX + MaxX) / 2
        MidY := (MinY + MaxY) / 2
        q.Q0.ApplyIfWithin(MinX, MinY, MidX, MidY, x, y, rx, ry, f)
        q.Q1.ApplyIfWithin(MidX, MinY, MaxX, MidY, x, y, rx, ry, f)
        q.Q2.ApplyIfWithin(MinX, MidY, MidX, MaxY, x, y, rx, ry, f)
        q.Q3.ApplyIfWithin(MidX, MidY, MaxX, MaxY, x, y, rx, ry, f)
    }
}

func (qt *QuadTree) ApplyIfWithin(x, y, rx, ry int, f func(paper *Paper)) {
    qt.Root.ApplyIfWithin(qt.MinX, qt.MinY, qt.MaxX, qt.MaxY, x, y, rx, ry, f)
}

func GenerateLabelZone(graph *Graph, scale, width, height, depth, xi, yi int, filename string) {

    if err := os.MkdirAll(filepath.Dir(filename),0755); err != nil {
        log.Fatal(err)
    }

    fo, _ := os.Create(filename+".json")
    defer fo.Close()
    w := bufio.NewWriter(fo)

    // Get midpoint of zone
    rx := width/2
    ry := height/2
    x  := graph.MinX + (xi-1)*int(width) + rx
    y  := graph.MinY + (yi-1)*int(height) + ry

    // TODO consider adding depth, x, y, width, height etc.
    // Tho in practice should already have this info before d/l label zone
    fmt.Fprintf(w,"lz_%d_%d_%d({\"scale\":%d,\"lbls\":[",depth,xi,yi,scale)

    min_rad := int(float64(scale)*0.01)

    first := true
    graph.qt.ApplyIfWithin(x, y, rx, ry, func(paper *Paper) {
        if paper.label != "" && paper.radius > min_rad {
            if first {
                first = false
            } else {
                fmt.Fprintf(w,",")
            }
            // TODO hopefully temporary
            label := cleanJsonString(paper.label)
            fmt.Fprintf(w,"{\"x\":%d,\"y\":%d,\"r\":%d,\"lbl\":\"%s\"}",paper.x,paper.y,paper.radius,label)
        }
    })

    fmt.Fprintf(w,"]})")
    w.Flush()
}

func ParallelGenerateLabelZone(graph *Graph, outPrefix string, depth, scale, width, height, xiFirst, xiLast, yiFirst, yiLast int, channel chan int) {
    for xi := xiFirst; xi <= xiLast; xi++ {
        for yi := yiFirst; yi <= yiLast; yi++ {
            filename := fmt.Sprintf("%s/zones/%d/%d/%d", outPrefix, depth, xi, yi)
            GenerateLabelZone(graph, scale, width, height, depth, xi, yi,filename)
        }
    }
    channel <- 1 // signal that this set of tiles is done
}

func GenerateAllLabelZones(graph *Graph, outPrefix string) {
    indexFile := outPrefix + "/zones/label_index.json"
    if err := os.MkdirAll(filepath.Dir(indexFile),0755); err != nil {
        fmt.Println(err)
        return
    }
    fo, _ := os.Create(indexFile)
    defer fo.Close()
    w := bufio.NewWriter(fo)

    sort.Sort(PaperSortId(graph.papers))
    latestId := graph.papers[0].id

    fmt.Fprintf(w,"label_index({\"latestid\":%d,\"xmin\":%d,\"ymin\":%d,\"xmax\":%d,\"ymax\":%d,\"zones\":[",latestId,graph.MinX,graph.MinY,graph.MaxX,graph.MaxY,)

    // tile divisions, scale divisions
    depthSet := []struct {
        tdivs, sdivs uint  
    }{
        {1,1},
        {1,2},
        {1,4},
        {1,8},
        {2,16},
        {4,32},
        {8,64},
        {16,128},
        {32,256},
    }

    first := true

    for depth, labelDepth := range depthSet {
        //divs := int(math.Pow(2.,float64(depth)))
        tile_width := int(math.Max(float64(graph.BoundsX)/float64(labelDepth.tdivs), float64(graph.BoundsY)/float64(labelDepth.tdivs)))
        tile_height := tile_width

        // typical scale of tile
        scale := int(math.Max(float64(graph.BoundsX)/float64(labelDepth.sdivs), float64(graph.BoundsY)/float64(labelDepth.sdivs)))

        if !first {
             fmt.Fprintf(w,",")
        }
        first = false
        fmt.Fprintf(w,"{\"z\":%d,\"s\":%d,\"w\":%d,\"h\":%d,\"nx\":%d,\"ny\":%d}",depth, scale, tile_width, tile_height,labelDepth.tdivs,labelDepth.tdivs)

        if !*flagSkipZones {
            fmt.Printf("Generating label zones at depth %d\n",depth)
            // TODO if graph far from from square, shorten tile directions accordingly

            // parallelise the drawing of zones, using as many cpus as we have available to us
            maxCpu := runtime.NumCPU()
            if *flagMaxCores > 0 && *flagMaxCores < maxCpu {
                maxCpu = *flagMaxCores
            }
            runtime.GOMAXPROCS(maxCpu)
            channel := make(chan int, maxCpu)
            numRoutines := 0
            xiPerCpu := (int(labelDepth.tdivs) + maxCpu - 1) / maxCpu
            for xi := 1; xi <= int(labelDepth.tdivs); {
                xiLast := xi + xiPerCpu - 1
                if xiLast > int(labelDepth.tdivs) {
                    xiLast = int(labelDepth.tdivs)
                }
                go ParallelGenerateLabelZone(graph, outPrefix, depth, scale, tile_width, tile_height, xi, xiLast, 1, int(labelDepth.tdivs), channel)
                numRoutines += 1
                xi = xiLast + 1
            }
            // drain the channel
            for i := 0; i < numRoutines; i++ {
                <-channel // wait for one task to complete
            }
            // all tasks are finished
        }
    }
    fmt.Fprintf(w,"]})")
    w.Flush()
}
